### DISCUSSION TITLE
Form Editor... another sneak peek

### DISCUSSION THREAD

**Discussion by WaynePhillipsEA** [twinbasic/twinbasic#270] [General]
We now have a toolbox, property sheet, event list, copy & paste, and alignment/sizing.  Still a lot to do, but getting through it!

https://user-images.githubusercontent.com/25738109/124790283-f10e1680-df42-11eb-8598-6e025f078732.mp4

**Comment by bclothier** [comment]
Nice!

Will those implement various `IOle***` interfaces and some private VB interfaces or are we building it from grounds up?

**Comment by Kr00l** [comment]
Good to see GUI starting to evolve.

Will this be the cross-platform compatible GUI form ? Or the win32 compatible GUI form ?
Also the Events list states OnClick, OnDoubleClick etc. Will it be then in the code just Click, DblClick ?
e.g. (VB6 compatible)
```
Private Sub Command1_Click()
End Sub
```

**Comment by bclothier** [comment]
One more follow up. A design I didn't like too much about VBx was that the definitions for the forms were basically hidden and not easily accessible for text editing. 

In .NET, the solution was to create a `.designer` file which provide all the additional code that was normally done by GUI design (e.g. creating & placing a control). I'm kind of lukewarm on that solution but it is a workable solution for keeping it in source code control. 

What approach will the `tB` take here? Will the `.tbf` file be a binary file or a text file?

**Comment by wqweto** [comment]
You might want to spend some time (or watch a tutorial) about Adobe designer products. I've been doing pre-press professionally (in my teens) and every "dev studio" compared to desktop publishing tools is completely inconvenient next to unusable.

No grouping (virtual groups, not common containers), manipulating z-order is nightmare, fine nudging, grid alignment, no keyboard shortcuts, zoom, quick alignment, quick duplicate -- nothing is done right. There is this attrib grid and you have to manually enter values which is not bad per se but it has to work very well for multi-selection too. Ctrl+Click, Alt+Click, Ctrl+Space+Click, right clicks, etc. modifiers not used at all.

Just blindly copying Illustrator features (or Photoshop) would be 100% better that present "dev studios" or inventing the wheel yourself IMO.

**Comment by WaynePhillipsEA** [comment]
For the initial offering, we're doing the cross platform versions, where the controls are drawn on a simple 2d canvas (32 bit with alpha).  That's where I'm hoping you guys can help out?!  I'm creating a framework for making these custom drawn controls inside tB, and then I'll see what you guys come up with whilst I move on to some of the other important things that are in the queue for tB.

The VB6 compatible forms can come a bit later, once the form designer has matured.

@Kr00l don't worry about the event names;. I will get them right when I actually implement them.

@bclothier the form designer works entirely with JSON documents, and that's how they're stored in the source file, only compiling down to binary in the EXE/DLL.

**Comment by WaynePhillipsEA** [comment]
@wqweto I totally appreciate your points, and I'll certainly be listening once the form designer is in your hands.  We certainly want to make something better than what's already out there, but we do have to start somewhere and then I can get feedback and make improvements along the way.

**Comment by Kr00l** [comment]
It's a difficult area for sure. Also I guess it's not straight forward to plan with and will go a long way.

For example one question I am curios.
Can a project have cross-platform and win32 form in one project? (Means the VB.Forms collection can present both?)
Or will there be exclusive depending on a setting?
(Both ways, exclusive or not, have pros and cons for sure)

**Comment by WaynePhillipsEA** [comment]
@Kr00l, yes mixing the form types won't be a problem.  Just as an aside, if I remember correctly, Microsoft Access forms are also custom drawn, though they do create 'real' Windows controls at times (for example when a text box gets the focus).

**Comment by WaynePhillipsEA** [comment]
Once we've got reasonable cross-platform forms support, there's really not much else to stop us moving on to full cross platform support in tB, and that's quite an exciting prospect.

**Comment by camomille22003** [comment]
If it helps to get some ideas, compare with my other favorite IDE (VisualFreeBasic)

![Image1](https://user-images.githubusercontent.com/33807881/124874021-80f5a400-dfc7-11eb-95f5-76c2f7937836.jpg)

Lock control
Delete control
Left alignment
Vertical alignment
Right alignment
Top alignment
Horizontal alignment
Bottom end alignment
Size: same width
Size: same height
Size: same
Horizontal spacing: same spacing
Horizontal spacing: increment
Horizontal spacing: decrement
Horizontal spacing: removal interval
Vertical spacing: same spacing
Vertical spacing: increment
Vertical spacing: decrement
Vertical spacing: removal interval
In the form: horizontal
In the form: vertical
In the form: midle
Order: Move to the top
Order: Move to the bottom

**Comment by WaynePhillipsEA** [comment]
I'm thinking of allowing our cross-platform forms to have different 'layouts' where each layout would be selectable by a tab just above the form designer.  This would allow us to design different layouts of the form controls.   For example, we might have a 'Desktop Layout' and a 'Tablet Layout' for a form, where both layouts would have the exact same controls in them, but the sizes/layout/visibility might differ significantly.   The decision of which layout to use could be made at compile time (through compiler conditionals for example when building for a Tablet build configuration), or at runtime (e.g. decided by the screen size).

I appreciate there is much crossover with features like anchoring, and even things like MVP/MVVM design patterns, so I just wanted to throw the idea out to you guys first to gauge your responses.

Any thoughts?

**Reply by mansellan** [reply]
Hmm, not sure that's optimal. With anchoring, docking and a few container types (stackpanel, grid, etc), it should be possible to implement a fully responsive design.

That said, it would definitely be useful to be able to preview at different aspects, as one can with web development tools.

Just my 2c.

**Reply by WaynePhillipsEA** [reply]
Thanks @mansellan.. I like the preview idea too

**Reply by mwolfe02** [reply]
I like the idea of multiple layouts as an option.  I think of it as being analogous to CSS Media queries.  I do not think that should be done instead of anchoring and docking, but rather as an additional option (assuming it's not a tremendous amount of additional work on your end).

**Comment by bimimicah** [comment]
Will you use the UxTheme APIs on Windows to provide OS-native visual styles? That will be a significant improvement over the visuals I see here, though I suppose this is still alpha and subject to change.
Looks like a very promising project!
